lib/matrix.js

```diff
@@ -52,11 +58,351 @@ var init = function(exports){
+        //   event dup suppression? - apparently we should still be doing this
+        //   tracking current display name / avatar per-message
+        //   pagination
+        //   re-sending (including persisting pending messages to be sent)
+        //   - Need a nice way to callback the app for arbitrary events like displayname changes
+        //   due to ambiguity (or should this be on a chat-specific layer)?
+                        
+        /*
+         * Helper method for retrieving the name of a room suitable for display in the UI
+         * TODO: in future, this should be being generated serverside.
+         */
+        getFriendlyRoomName: function(roomId) {
+            // we need a store to track the inputs for calculating room names
+            if (!this.store) return roomId;
+            
+            // check for an alias, if any. for now, assume first alias is the official one.
+            var alias;
+            var mRoomAliases = this.store.getStateEvent(roomId, 'm.room.aliases');
+            if (mRoomAliases) {
+                alias = mRoomAliases.event.content.aliases[0];
+            }
+            
+            var mRoomName = this.store.getStateEvent(roomId, 'm.room.name');
+            if (mRoomName) {
+                return mRoomName.event.content.name + (alias ? " (" + alias + ")": "");
+            }
+            else if (alias) {
+                return alias;
+            }
+            else {
+                var userId = this.credentials.userId;
+                var members = this.store.getStateEvents(roomId, 'm.room.member')
+                    .filter(function(event) {
+                        return event.event.user_id !== userId;
+                    });
+                
+                if (members.length == 0) {
+                    return "Unknown";
+                }
+                else if (members.length == 1) {
+                    return members[0].event.content.displayname || members[0].event.user_id;
+                }
+                else if (members.length == 2) {
+                    return (members[0].event.content.displayname || members[0].event.user_id) + " and " +
+                           (members[1].event.content.displayname || members[1].event.user_id);
+                }
+                else {
+                    return (members[0].event.content.displayname || members[0].event.user_id) + " and " +
+                           (members.length - 1) + " others";
+                }
+            }
+        },
+        
+        /*
+         * Helper method for retrieving the name of a user suitable for display in the UI
+         * in the context of a room - i.e. disambiguating from any other users in the room.
+         * XXX: This could perhaps also be generated serverside, perhaps by just passing 
+         * a 'disambiguate' flag down on membership entries which have ambiguous displaynames?
+         */
+        getFriendlyDisplayName: function(userId, roomId) {
```

Comment:
I feel the `getFriendly` and co. functions are in the right(ish) place but not separated enough from the `MatrixClient`. See main comment.